Add phone dialer component to the action sidebar

nmajor25 opened the pull request:
Shortcut Story ID: [sc-63637]

- [x] Implement in app UI
- [x] Disabled when action is in read only state
- [x] Cypress tests
- [x] Code cleanup/refactoring

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Patient sidebar adds a Dialer for quick access to patient phone numbers and one‑click calling (when enabled); new dialer component and view integrated.

- **UI**
  - Picklist shows an animated loading state with an icon; sidebar has a dedicated dialer area.

- **Style**
  - New styles for dialer button, labels, highlighted states, and loading animation.

- **Localization**
  - Added English dialer labels and texts.

- **Tests**
  - Integration specs covering the dialer workflow and UI interactions.

- **Bug Fixes**
  - Option list selection now safely handles missing selection callbacks; added two new icons: circle and phone.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

paulfalgout committed:
Introduce dialer component

nmajor25 committed:
Add icon and animation to picklist isLoading view

nmajor25 commented on src/js/components/picklist/picklist.scss:
```diff
@@ -67,8 +67,18 @@
   padding: 4px 16px;
+}
 
-  .icon:first-child {
```
I couldn't find anywhere where this is used. No results text that has an icon as well.

nmajor25 commented on src/js/components/picklist/index.js:
```diff
@@ -53,14 +54,27 @@ const PicklistEmpty = View.extend({
-      {{ loadingText }}
+      <div class="picklist__message-loading js-loading">
+        {{fas "circle"}}{{ loadingText }}
+      </div>
```
This makes this the loading ui/styling for any picklist with the `isListsAsync: true` option. I figured this is fine for now. And we can figure out a way to customize via an options item it in the future if the need arises. Like if we want to use a different icon, spacing, or something across different picklists.

nmajor25 replied:
Other way would be to pass a `loadingTemplate`option or something similar that's a handlebars template.

paulfalgout commented on src/js/views/patients/shared/components/dialer-component.scss:
```diff
@@ -0,0 +1,36 @@
+.dialer-component__phone-icon {
+  color: #0582DA;
```
color vars

paulfalgout commented on src/js/views/patients/shared/components/dialer_component.js:
```diff
@@ -0,0 +1,96 @@
+  viewOptions() {
+    return {
+      tagName: 'button',
+      className: 'button-secondary w-100 action-sidebar__form',
```
should probably make action-sidebar__form something more generic if it should be applied here as well

nmajor25 replied:
We might not need that. Looks like it's just applying the blue coloring. 

I'm going to go clean some of the code up in a final commit. I'll include this change.

paulfalgout commented on src/js/views/patients/shared/components/dialer_component.js:
```diff
@@ -0,0 +1,96 @@
+    const lists = this.getLists();
+
+    const optionlist = new Optionlist({
+      attr: 'number',
```
probably don't need this now

nmajor25 committed:
Disable dialer button if action is done or action.canEdit = false

nmajor25 committed:
Add cy.tick() to beginning of 'display action sidebar' test

To fix issue where sidebar appeared empty in test replays, even though test was passing.

Since this test uses visit.onClock(), its needed for the 300ms animation when the sidebar is opened

nmajor25 committed:
Add 'dialer' setting to test fixtures, with a 'null' value by default

nmajor25 committed:
Add cypress tests for dialer button in action sidebar

nmajor25 commented on test/integration/patients/sidebar/action-sidebar.js:
```diff
@@ -22,7 +22,6 @@ import { roleNoFilterEmployee, roleTeamEmployee } from 'support/api/roles';
 import { getFlow } from 'support/api/flows';
 import { getFile } from 'support/api/files';
 
-// TODO: Update to mergejson api
```
Don't think we need this comment anymore. These have all been switched over.

nmajor25 commented on test/integration/patients/sidebar/action-sidebar.js:
```diff
@@ -628,6 +629,11 @@ context('action sidebar', function() {
+    cy
+      .get('.sidebar')
+      .find('[data-dialer-region]')
+      .should('be.empty');
```
Ensures dialer is not shown when 'settings.dialer' is not present or null.

nmajor25 commented on test/integration/patients/sidebar/action-sidebar.js:
```diff
@@ -712,6 +718,158 @@ context('action sidebar', function() {
       .should('contain', 'The Action you requested does not exist');
   });
 
+  specify('action phone dialer', function() {
```
Tests everything up until a phone number is clicked, which will open the five9 dialer UI when that's implemented.

Quick summary of test:

1. Open dialer picklist and verify phone numbers are shown properly and are ordered based on `preferred` value.
2. Close dialer picklist and open it again. Verify phone numbers are displayed via cached data. Api request shouldn't be made again.
3. Change action to done and verify dialer button is set to `disabled`.
4. Change parent flow to done and verify dialer button is set to `disabled`.

nmajor25 replied:
Potentially move some of this to a component test on the `dialer_component.js`.

paulfalgout replied:
yep we don't need to test anything specific to the component here

paulfalgout commented on src/js/views/patients/shared/components/dialer_component.js:
```diff
@@ -0,0 +1,101 @@
+      },
+    });
+
+    phones.each(model => {
```
I think I'd like to move away from setting onSelect on the model.  at least in this case it makes less sense because every handler is the same.  

We maybe need to modify the optionlist to only call onSelect if it's set, but we should just be able to listen to the `select` event on the optionlist with the same handler

paulfalgout commented on src/js/components/picklist/index.js:
```diff
@@ -53,14 +54,27 @@ const PicklistEmpty = View.extend({
+      ease: 'inOutSine',
+      alternate: true,
+    });
+  },
```
This is true.  animations that end effectively clean up themselves, but not looping animations.

paulfalgout commented on src/js/views/patients/shared/components/dialer_component.js:
```diff
@@ -0,0 +1,101 @@
+
+    phones.each(model => {
+      model.set('onSelect', () => {
+        Radio.request('dialer', 'call', model.get('number'));
```
ah one other thing the 4th argument for this request needs to be the action model @nmajor25

nmajor25 committed:
Use color variables in dialer-component.scss file

nmajor25 committed:
Use more appropriate naming for form/dialer button blue text class

nmajor25 committed:
Remove 'attr: 'number'' from dialer options list options

nmajor25 committed:
Pass action model as param in 'dialer' => 'call' radio request

nmajor25 committed:
Listen to select events on dialer optionlist

Instead of setting onSelect's on each list item model

In option list component, handle when a list item is clicked but has no onSelect

nmajor25 committed:
In picklist only show loader animation if this.ui.loading[0] exists. And clean up animation on destroy

paulfalgout commented on src/js/components/picklist/index.js:
```diff
@@ -53,14 +54,35 @@ const PicklistEmpty = View.extend({
     return this.options;
   },
+  onRender() {
+    if (this.ui.loading[0]) {
```
I think we should check the data and not the dom for this check

nmajor25 replied:
This is updated to check `this.getOption('isLoading')` instead of the dom.

nmajor25 replied:
Although I do think this would be simpler not using animejs and just doing it in the CSS instead.

Via:

```css
.picklist__message-loading {
  animation: pulse 400ms ease-in-out infinite alternate;

  . . .
}

@keyframes pulse {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}
```

nmajor25 replied:
But we can look into that in a future task if we want to do that instead.

paulfalgout replied:
css animation seems fine to me.. cleaner

nmajor25 replied:
I just switched this over to handle the animation in css instead of animejs: https://github.com/RoundingWell/care-ops-frontend/pull/1522/commits/f2ad84769485f80e99b94eb73ae58ce088c6a755

paulfalgout commented on src/js/views/patients/shared/components/dialer-component.scss:
```diff
@@ -0,0 +1,36 @@
+    align-items: center;
+    display: none;
+
+    .icon {
```
this icon styles look very unique.  don't see 11px anywhere and don't love the 2px either..

nmajor25 replied:
Updated font-size to `12px`

nmajor25 replied:
The `2px` makes the alignment appear a little better, is why I added it. The icon is vertically centered via flex and align items center. But the design of the icon next to the text makes it look off (at least to me).

With `margin-bottom: 2px`:

<img width="141" height="72" alt="Screenshot 2025-09-30 at 11 28 50 AM" src="https://github.com/user-attachments/assets/53b14a7f-edb4-4ca9-888b-1d3d74e26156" />

Without:

<img width="113" height="72" alt="Screenshot 2025-09-30 at 11 29 00 AM" src="https://github.com/user-attachments/assets/94976370-d372-4d28-921d-326cbbd74dac" />

paulfalgout replied:
does vertical-align on the icon itself do anything?  might not with line-height.. but like baseline or middle might improve it as well.

nmajor25 replied:
Vertical align doesn't seem to have any effect...

paulfalgout replied:
well we should at least add a comment so I don't remove it later

paulfalgout commented on src/js/views/patients/sidebar/action/action-sidebar.scss:
```diff
@@ -5,7 +5,8 @@
 
-.action-sidebar__form {
+.action-sidebar__form-button,
+.action-sidebar__dialer-button {
```
I don't think it being blue is related to form or dialer.. this should be something like action-sidebar__function

nmajor25 replied:
~Updated this to one `.action-sidebar__blue-text` class~

nmajor25 replied:
Updated this to one `.action-sidebar__button` class: https://github.com/RoundingWell/care-ops-frontend/pull/1522/commits/7dba12e9dadf79d88736cb82d91c2bf87b4eb652

nmajor25 committed:
Change .action-sidebar__form-button/.action-sidebar__dialer-button => .action-sidebar__button

nmajor25 committed:
Update .dialer-component__phone-label-call => .icon to 'font-size: 12px'

nmajor25 committed:
For picklist loading msg animation, use css instead of animejs

paulfalgout commented on src/js/components/picklist/picklist.scss:
```diff
@@ -67,8 +67,29 @@
+}
 
-  .icon:first-child {
+@keyframes pulse {
+  from {
+    opacity: 0;
+  }
+
+  to {
+    opacity: 1;
+  }
+}
```
@nmajor25 what do you think about this.. maybe .picklist prefixing the keyframes is a good idea to prevent collisions

nmajor25 replied:
It wouldn't hurt 👍

shadowhand replied:
The 🐇 did good with that one. Especially like the accessibility recommendation for reduced animation.

paulfalgout commented on src/js/components/picklist/index.js:
```diff
@@ -53,11 +53,16 @@ const PicklistEmpty = View.extend({
       {{ noResultsText }}
     {{/if}}
   `,
+  ui: {
```
looks like we can remove the selector and ui element?

nmajor25 replied:
Yep, not needed anymore

nmajor25 replied:
That's been removed now 👍

paulfalgout commented on src/js/components/picklist/index.js:
```diff
@@ -53,11 +53,16 @@ const PicklistEmpty = View.extend({
+        {{fas "circle"}}{{ loadingText }}
+      </div>
     {{else}}
       {{ noResultsText }}
```
what if we wrapped this in a div class="picklist__message"  and removed it from the view className?

nmajor25 replied:
That works, updated in https://github.com/RoundingWell/care-ops-frontend/pull/1522/commits/4ebbc7f44da91e3106c38450c25409560014a497.

nmajor25 replied:
But broke some tests...

paulfalgout reviewed (changes requested):
just down to nitpicks I think

shadowhand commented on src/js/components/picklist/index.js:
```diff
@@ -53,11 +53,16 @@ const PicklistEmpty = View.extend({
   template: hbs`
     {{#if isLoading}}
-      {{ loadingText }}
+      <div class="picklist__message-loading js-loading">
```
Is `js-loading` still needed, now that the animation is in CSS?

nmajor25 replied:
Nope, not needed 👍

nmajor25 committed:
Move .picklist__message off the view class and onto the {{noResultsText}}

paulfalgout reviewed (approved):
Looks good. can refactor later if necessary
